File: src/main/java/io/scriptor/fdt/BuilderContext.java
package io.scriptor.fdt;

import com.carrotsearch.hppc.ObjectIntHashMap;
import com.carrotsearch.hppc.ObjectIntMap;
import org.jetbrains.annotations.NotNull;

public final class BuilderContext<T> {

    private final ObjectIntMap<T> handles = new ObjectIntHashMap<>();

    public int get(final @NotNull T key) {
        if (handles.containsKey(key))
            return handles.get(key);

        final var handle = handles.size() + 1;
        handles.put(key, handle);
        return handle;
    }
}

File: src/main/java/io/scriptor/fdt/BuilderContext.kt
package io.scriptor.fdt

class BuilderContext<T> {

    private val handles: MutableMap<T, UInt> = HashMap()

    fun get(key: T): UInt {
        if (key in handles)
            return handles[key]!!

        val handle = handles.size.toUInt() + 1u
        handles[key] = handle
        return handle
    }
}
